xplique/attributions/opti_cam.py

```diff
@@ -0,0 +1,185 @@
+
+        Returns
+        -------
+        opti_cams
+            Opti-CAM explanations, same shape as the inputs except for the channels.
+        """
+        # pylint: disable=E1101
+        batch_size = self.batch_size if self.batch_size is not None else len(inputs)
+
+        # optimization loop is done for each batch separately
+        opti_cams = None
+        for x_batch, y_batch in tf.data.Dataset.from_tensor_slices((inputs, targets)).batch(
+                batch_size):
+            # initialize weights and optimization elements
+            current_batch_size = x_batch.shape[0]
+            weights = tf.Variable(
+                0.5 * tf.ones((current_batch_size, 1, 1, self.conv_layer.output.shape[-1])
+            ), trainable=True, dtype=tf.float32)
+            optimizer = tf.keras.optimizers.Adam(0.05)
+            for _ in range(self.n_iters):
+                with tf.GradientTape(watch_accessed_variables=False) as tape:
+                    tape.watch(weights)
+                    feature_maps, logits = self.model(x_batch)
+                    explanations = self._one_step_explanation(x_batch, feature_maps, weights)
+                    explanations = tf.map_fn(self.normalize, explanations)
+                    x_perturbed = tf.multiply(
+                        tf.tile(explanations, [1, 1, 1, x_batch.shape[-1]]),
+                        x_batch
+                    )
+                    _, logits_perturbed = self.model(x_perturbed)
+                    logits = tf.reduce_sum(logits * y_batch, axis=1)
+                    logits_perturbed = tf.reduce_sum(logits_perturbed * y_batch, axis=1)
+                    score = self.loss_fn(logits, logits_perturbed)
+                grads = tape.gradient(score, weights)
+                optimizer.apply_gradients([(grads, weights)])
+
+            # Generate the final explanations for the batch of images
+            explanations = self._one_step_explanation(x_batch, feature_maps, weights)
+            explanations = tf.map_fn(self.normalize, explanations)
+            opti_cams = explanations if opti_cams is None \
+                else tf.concat([opti_cams, explanations], axis=0)
+
+        return opti_cams
+
+    @staticmethod
+    @tf.function
+    def _one_step_explanation(images: tf.Tensor, feature_maps: tf.Tensor, weights: tf.Tensor):
+        """
+        Generate the explanation for a single step of the optimization.
+
+        Parameters
+        ----------
+        images
+            Original images that we wish to explain.
+        feature_maps
+            Feature maps of the convolutional layer for the images.
+        weights
+            Weights of the feature maps (aka the CAM explanation) for the images.
+
+        Returns
+        -------
+        overlayed_images
+            Images with the explanation overlayed.
+        """
+        alpha = tf.nn.softmax(weights, axis=-1)
+        weighted_explanation = tf.reshape(
```

Comment:
There should be a clearer way to write this. Maybe with `tf.einsum` or `tf.matvec`.